{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2025-10-27 19:07:09--  http://hgdownload.soe.ucsc.edu/goldenPath/hg38/chromosomes/chr10.fa.gz\n",
      "Resolving hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)... 128.114.119.163\n",
      "Connecting to hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)|128.114.119.163|:80... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 43157332 (41M) [application/x-gzip]\n",
      "Saving to: 'chr10.fa.gz'\n",
      "\n",
      "chr10.fa.gz         100%[===================>]  41.16M   945KB/s    in 2m 2s   \n",
      "\n",
      "2025-10-27 19:09:18 (346 KB/s) - 'chr10.fa.gz' saved [43157332/43157332]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget http://hgdownload.soe.ucsc.edu/goldenPath/hg38/chromosomes/chr10.fa.gz\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "08a98036",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "zsh:1: no matches found: [URL_TO_ILLUMINA_FILE]\n"
     ]
    }
   ],
   "source": [
    "!wget [URL_TO_ILLUMINA_FILE] -O illumina.fq.bz2\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "35323d22",
   "metadata": {},
   "outputs": [],
   "source": [
    "!bunzip2 -k illumina.fq.bz2\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2de07851",
   "metadata": {},
   "outputs": [],
   "source": [
    "!bunzip2 -k pacbio.fq.bz2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "02e427b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::mm_idx_gen::2.313*0.94] collected minimizers\n",
      "[M::mm_idx_gen::2.680*1.21] sorted minimizers\n",
      "[M::main::2.680*1.21] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::2.680*1.21] mid_occ = 1000\n",
      "[M::mm_idx_stat] kmer size: 21; skip: 11; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::2.809*1.20] distinct minimizers: 19023001 (96.65% are singletons); average occurrences: 1.171; average spacing: 6.004; total length: 133797422\n",
      "[M::worker_pipeline::10.715*2.40] mapped 309505 sequences\n",
      "[M::main] Version: 2.30-r1287\n",
      "[M::main] CMD: minimap2 -ax sr chr10.fa illumina.fq\n",
      "[M::main] Real time: 10.759 sec; CPU: 25.739 sec; Peak RSS: 0.882 GB\n"
     ]
    }
   ],
   "source": [
    "!minimap2 -ax sr chr10.fa illumina.fq | samtools sort -o illumina.bam\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a8dc871f",
   "metadata": {},
   "outputs": [],
   "source": [
    "!samtools index illumina.bam\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "69b55650",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::mm_idx_gen::1.967*0.94] collected minimizers\n",
      "[M::mm_idx_gen::2.247*1.18] sorted minimizers\n",
      "[M::main::2.247*1.18] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::2.418*1.15] mid_occ = 110\n",
      "[M::mm_idx_stat] kmer size: 19; skip: 19; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::2.477*1.14] distinct minimizers: 11183865 (96.18% are singletons); average occurrences: 1.195; average spacing: 10.008; total length: 133797422\n",
      "[M::worker_pipeline::3.540*1.64] mapped 3063 sequences\n",
      "[M::main] Version: 2.30-r1287\n",
      "[M::main] CMD: minimap2 -ax map-hifi chr10.fa pacbio.fq\n",
      "[M::main] Real time: 3.572 sec; CPU: 5.823 sec; Peak RSS: 0.755 GB\n"
     ]
    }
   ],
   "source": [
    "!minimap2 -ax map-hifi chr10.fa pacbio.fq | samtools sort -o pacbio.bam\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4c30be47",
   "metadata": {},
   "outputs": [],
   "source": [
    "!samtools index pacbio.bam\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "92c1e508",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r--@ 1 tomislavstojcic  staff    20M Oct 27 19:40 illumina.bam\n",
      "-rw-r--r--@ 1 tomislavstojcic  staff    79K Oct 27 19:40 illumina.bam.bai\n",
      "-rw-r--r--@ 1 tomislavstojcic  staff   2.7M Oct 27 19:40 pacbio.bam\n",
      "-rw-r--r--@ 1 tomislavstojcic  staff    58K Oct 27 19:40 pacbio.bam.bai\n"
     ]
    }
   ],
   "source": [
    "ls -lh *.bam *.bai\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ba2dd87e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Note: none of --samples-file, --ploidy or --ploidy-file given, assuming all sites are diploid\n",
      "[mpileup] 1 samples in 1 input files\n",
      "[mpileup] maximum number of reads per input file set to -d 250\n"
     ]
    }
   ],
   "source": [
    "!bcftools mpileup -f chr10.fa illumina.bam | bcftools call -mv -Oz -o illumina.vcf.gz\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6b3b4df0",
   "metadata": {},
   "outputs": [],
   "source": [
    "!bcftools index illumina.vcf.gz\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "cf46980d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Note: none of --samples-file, --ploidy or --ploidy-file given, assuming all sites are diploid\n",
      "[mpileup] 1 samples in 1 input files\n",
      "[mpileup] maximum number of reads per input file set to -d 250\n"
     ]
    }
   ],
   "source": [
    "!bcftools mpileup -f chr10.fa pacbio.bam | bcftools call -mv -Oz -o pacbio.vcf.gz\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f5b2b6b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "!bcftools index pacbio.vcf.gz\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "243ef250",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "zsh:1: command not found: whatshap\n"
     ]
    }
   ],
   "source": [
    "!whatshap phase -o illumina_phased.vcf --reference chr10.fa illumina.vcf.gz illumina.bam\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "922d566a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is WhatsHap 2.8 running under Python 3.11.5\n",
      "Working on 1 sample from 1 family\n",
      "WARNING: Skipping duplicated position 94664912 on chromosome 'chr10' Hiding further warnings of this type, use --debug to show\n",
      "\n",
      "# Working on contig chr10 in individual illumina.bam\n",
      "Found 1301 usable heterozygous variants (0 skipped due to missing genotypes)\n",
      "WARNING: Sample 'illumina.bam' not found in any BAM/CRAM file.\n",
      "Found 0 reads covering 0 variants\n",
      "Kept 0 reads that cover at least two variants each\n",
      "Selected 0 most phase-informative reads covering 0 variants\n",
      "Best-case phasing would result in 0 non-singleton phased blocks (0 singletons). \n",
      "Phasing 1 sample by solving the MEC problem ...\n",
      "No. of phased blocks: 0\n",
      "\n",
      "# Resource usage\n",
      "Time spent reading BAM/CRAM:                    0.3 s\n",
      "Time spent parsing VCF:                         0.0 s\n",
      "Time spent selecting reads:                     0.0 s\n",
      "Time spent phasing:                             0.0 s\n",
      "Time spent writing VCF:                         0.0 s\n",
      "Time spent finding components:                  0.0 s\n",
      "Time spent on rest:                             0.0 s\n",
      "Total elapsed time:                             0.3 s\n"
     ]
    }
   ],
   "source": [
    "!whatshap phase -o illumina_phased.vcf --reference chr10.fa illumina.vcf.gz illumina.bam\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "fd81ede3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is WhatsHap 2.8 running under Python 3.11.5\n",
      "Working on 1 sample from 1 family\n",
      "\n",
      "# Working on contig chr10 in individual pacbio.bam\n",
      "Found 352 usable heterozygous variants (0 skipped due to missing genotypes)\n",
      "WARNING: Sample 'pacbio.bam' not found in any BAM/CRAM file.\n",
      "Found 0 reads covering 0 variants\n",
      "Kept 0 reads that cover at least two variants each\n",
      "Selected 0 most phase-informative reads covering 0 variants\n",
      "Best-case phasing would result in 0 non-singleton phased blocks (0 singletons). \n",
      "Phasing 1 sample by solving the MEC problem ...\n",
      "No. of phased blocks: 0\n",
      "\n",
      "# Resource usage\n",
      "Time spent reading BAM/CRAM:                    0.3 s\n",
      "Time spent parsing VCF:                         0.0 s\n",
      "Time spent selecting reads:                     0.0 s\n",
      "Time spent phasing:                             0.0 s\n",
      "Time spent writing VCF:                         0.0 s\n",
      "Time spent finding components:                  0.0 s\n",
      "Time spent on rest:                             0.0 s\n",
      "Total elapsed time:                             0.3 s\n"
     ]
    }
   ],
   "source": [
    "!whatshap phase -o pacbio_phased.vcf --reference chr10.fa pacbio.vcf.gz pacbio.bam\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ba07e560",
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip illumina_phased.vcf\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "0560e514",
   "metadata": {},
   "outputs": [],
   "source": [
    "!bcftools index illumina_phased.vcf.gz\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "fe746ce5",
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip pacbio_phased.vcf\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b5d9487b",
   "metadata": {},
   "outputs": [],
   "source": [
    "!bcftools index pacbio_phased.vcf.gz\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "42882f20",
   "metadata": {},
   "outputs": [],
   "source": [
    "!bcftools isec -p isec_output illumina_phased.vcf.gz pacbio_phased.vcf.gz\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "eeacd76c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Illumina-only variants:\n"
     ]
    }
   ],
   "source": [
    "!echo \"Illumina-only variants:\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "637d1925",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    1437 isec_output/0000.vcf\n"
     ]
    }
   ],
   "source": [
    "!wc -l isec_output/0000.vcf\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "da58b409",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PacBio-only variants:\n"
     ]
    }
   ],
   "source": [
    "!echo \"PacBio-only variants:\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "d0c0f05f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     264 isec_output/0001.vcf\n"
     ]
    }
   ],
   "source": [
    "!wc -l isec_output/0001.vcf\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "26d4e624",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shared variants:\n"
     ]
    }
   ],
   "source": [
    "!echo \"Shared variants:\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "6048305f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     569 isec_output/0002.vcf\n"
     ]
    }
   ],
   "source": [
    "!wc -l isec_output/0002.vcf\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "df00c26c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CYP2C19 variant count:\n",
      "     125\n"
     ]
    }
   ],
   "source": [
    "print(\"CYP2C19 variant count:\")\n",
    "!bcftools view -H -r chr10:94759681-94853205 illumina_cyp_phased.vcf.gz | wc -l\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f708406f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "CYP2C9 variant count:\n",
      "      71\n"
     ]
    }
   ],
   "source": [
    "print(\"\\nCYP2C9 variant count:\")\n",
    "!bcftools view -H -r chr10:94938683-94993415 illumina_cyp_phased.vcf.gz | wc -l\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "dedbc320",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "CYP2C8 variant count:\n",
      "     129\n"
     ]
    }
   ],
   "source": [
    "print(\"\\nCYP2C8 variant count:\")\n",
    "!bcftools view -H -r chr10:94842701-94924000 illumina_cyp_phased.vcf.gz | wc -l\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a6e6b6c",
   "metadata": {},
   "source": [
    "## IGV Analysis of Discordant Variants\n",
    "\n",
    "### Variant 1: CYP2C19 chr10:94,772,788 \n",
    "![IGV Screenshot 1](Screenshot%202025-11-02%20at%207.34.32%20PM.png)\n",
    "\n",
    "**Technology:** Illumina only  \n",
    "**Variant:**  \n",
    "**Analysis:** This variant shows strong Illumina support with good coverage. PacBio has no reads in this region, suggesting a coverage gap rather than a sequencing artifact. This is likely a real variant.\n",
    "\n",
    "---\n",
    "\n",
    "### Variant 2: CYP2C19 chr10:94,770,084\n",
    "![IGV Screenshot 2](Screenshot%202025-11-02%20at%207.38.35%20PM.png)\n",
    "\n",
    "**Coverage:** Illumina: 35X, PacBio: 34X  \n",
    "**Analysis:** Illumina shows a large indel (blue/green colored block) with strong support across multiple reads. PacBio has adequate coverage but shows only minimal variant support. The PacBio reads appear to disagree on the exact nature of this variant. Most likely a real indel that Illumina captures clearly but PacBio has difficulty resolving, possibly due to alignment challenges in this region.\n",
    "\n",
    "---\n",
    "\n",
    "### Variant 3: CYP2C9 chr10:94,969,284 \n",
    "![IGV Screenshot 3](Screenshot%202025-11-02%20at%207.39.33%20PM.png)\n",
    "**Coverage:** Illumina: 51X, PacBio: 37X  \n",
    "**Analysis:** PacBio shows 3 blue C's stacked at this position, representing only 3 out of 37 reads supporting this variant (8% allele frequency). Illumina data with 51X coverage shows no variant at this position. Thus this is probably not a true variant.\n",
    "\n",
    "---\n",
    "\n",
    "### Variant 4: CYP2C9 chr10:94,969,284 \n",
    "![IGV Screenshot 4](Screenshot%202025-11-02%20at%207.39.49%20PM.png)\n",
    "\n",
    "**Coverage:** Illumina: 43X, PacBio: 1X  \n",
    "**Analysis:** Illumina shows a large indel with strong support across 43 reads. PacBio has only 1 read in this region. The indel appears real based on consistent Illumina support, but cannot be validated due to insufficient PacBio coverage. Thus Likely a true variant; discordance due to PacBio coverage gap, not sequencing error.\n",
    "\n",
    "---\n",
    "\n",
    "### Variant 5: CYP2C8 chr10:94,847,100 \n",
    "![IGV Screenshot 5](Screenshot%202025-11-02%20at%207.40.17%20PM.png)\n",
    "\n",
    "**Coverage:** Illumina: 55X, PacBio: 26X  \n",
    "**Analysis:** Illumina shows multiple SNVs with strong read support across the region. PacBio has good coverage but shows minimal variant calls. Illumina variants likely true; PacBio single-read variant likely **artifact**. Discordance due to **variant calling sensitivity differences** between technologies, not sequencing errors."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8d973be2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=== CYP2C19 variants ===\n",
      "94760573\tC\tA\t0/1\n",
      "94761900\tC\tT\t0/1\n",
      "94762804\tC\tT\t1/1\n",
      "94763232\tA\tT\t0/1\n",
      "94765362\tG\tA\t0/1\n",
      "94765779\tG\tA\t1/1\n",
      "94766304\tttg\ttTGTAtg\t1/1\n",
      "94766352\tG\tA\t0/1\n",
      "94767381\tA\tC\t1/1\n",
      "94767577\tC\tA\t0/1\n",
      "94767965\tC\tG\t1/1\n",
      "94768394\tG\tA\t0/1\n",
      "94770155\tA\tG\t1/1\n",
      "94772788\tG\tT\t1/1\n",
      "94772850\tT\tC\t1|0\n",
      "94772907\tG\tA\t1|0\n",
      "94772931\tG\tA\t0|1\n",
      "94773293\tA\tT\t1/1\n",
      "94774827\tG\tA\t1/1\n",
      "94775961\tG\tA\t0/1\n",
      "94776698\tT\tA\t0|1\n",
      "94776951\tC\tT\t0|1\n",
      "94777632\tA\tG\t0/1\n",
      "94778957\tT\tA\t1|0\n",
      "94778981\tG\tA\t0|1\n",
      "94779551\ttttttcttttcttttctttt\ttTTTTCTTTTCttttcttttcttttctttt\t0|1\n",
      "94782837\tC\tT\t1|0\n",
      "94783106\tA\tG\t1|0\n",
      "94784235\tcttttt\tcTttttt\t1/1\n",
      "94784708\tG\tA\t1|0\n",
      "94784970\tC\tG\t1|0\n",
      "94785981\tC\tT\t0/1\n",
      "94787917\tG\tA\t0/1\n",
      "94788106\tgttccattcca\tgttcca\t0|1\n",
      "94788559\tT\tC\t0|1\n",
      "94789209\tA\tG\t0|1\n",
      "94791495\tG\tT\t0/1\n",
      "94792552\tG\tT\t0/1\n",
      "94793025\tA\tC\t1/1\n",
      "94794531\tA\tT\t1/1\n",
      "94794659\tA\tG\t1/1\n",
      "94795304\tT\tG\t0|1\n",
      "94795486\tC\tA\t0|1\n",
      "94795832\tC\tG\t0|1\n",
      "94797351\tT\tC\t0/1\n",
      "94798522\tC\tG\t1|0\n",
      "94798814\tatttttttttttttttttttttttt\tatttttttttttttttttttttt\t0|1\n",
      "94801232\tC\tT\t0|1\n",
      "94801768\tA\tG\t0|1\n",
      "94801966\tC\tA\t0|1\n",
      "94803367\tC\tT\t0/1\n",
      "94804815\tG\tT\t1/1\n",
      "94805643\tG\tA\t0/1\n",
      "94806641\ttatAATAAT\ttATAAT\t0/1\n",
      "94808278\tA\tT\t1/1\n",
      "94808754\tA\tC\t0|1\n",
      "94809169\tG\tA\t1|0\n",
      "94809696\tG\tC\t1|0\n",
      "94811053\tG\tA\t0/1\n",
      "94811953\tgttttttt\tgTttttttt\t0/1\n",
      "94812987\tT\tC\t0/1\n",
      "94813270\tA\tG\t1/1\n",
      "94814902\tT\tC\t0|1\n",
      "94815312\tC\tG\t0|1\n",
      "94816092\tATTTTTTTT\tATTTTTTT\t0|1\n",
      "94816244\tcttttttttttt\tctttttttttt\t0|1\n",
      "94816722\ttcccccc\ttccccc\t0|1\n",
      "94817128\tA\tG\t1|0\n",
      "94817208\tG\tA\t1|0\n",
      "94817354\tT\tA\t1|0\n",
      "94817647\tT\tC\t1/1\n",
      "94817752\tA\tG\t1/1\n",
      "94818012\tcaaaaaaaaaaaaaaaaaaaaaaaa\tcaaaaaaaaaaaaaaaaaaaaaa\t1/1\n",
      "94818356\tT\tA\t1|0\n",
      "94818364\tA\tG\t1|0\n",
      "94819214\tA\tT\t0|1\n",
      "94819415\tC\tT\t1|0\n",
      "94822035\tG\tT\t1/1\n",
      "94822399\tA\tC\t0/1\n",
      "94823871\tT\tC\t1/1\n",
      "94824667\tT\tA\t0/1\n",
      "94827531\tC\tT\t0|1\n",
      "94827532\tG\tA\t1|0\n",
      "94827790\tT\tG\t0|1\n",
      "94827874\tT\tC\t1/1\n",
      "94828624\tA\tG\t0/1\n",
      "94829705\tC\tT\t0|1\n",
      "94829819\tG\tA\t0|1\n",
      "94830368\tT\tC\t0|1\n",
      "94830377\tA\tT\t0|1\n",
      "94830470\tC\tT\t0|1\n",
      "94830487\tT\tG\t1|0\n",
      "94831197\tA\tG\t0/1\n",
      "94833462\tT\tA\t0/1\n",
      "94834134\tG\tT\t0|1\n",
      "94834676\tA\tC\t0|1\n",
      "94836209\tC\tT\t1/1\n",
      "94836383\tG\tA\t0/1\n",
      "94836408\tT\tA\t1/1\n",
      "94837046\tgaaaaa\tgAaaaaa\t0/1\n",
      "94837994\tA\tC\t0/1\n",
      "94839520\tA\tG\t0|1\n",
      "94839753\tG\tA\t0|1\n",
      "94839916\tG\tA\t1/1\n",
      "94839932\tC\tA\t1/1\n",
      "94839990\tattttttttt\taTttttttttt\t0|1\n",
      "94840228\tcttttttttttttt\tctttttttttttt\t0|1\n",
      "94841025\tG\tT\t0|1\n",
      "94841089\tG\tA\t0|1\n",
      "94841819\tC\tT\t1/1\n",
      "94842037\tA\tG\t0/1\n",
      "94842393\tgtttttttttttttttttttt\tgttttttttttttttttttt,gTtttttttttttttttttttt\t1/2\n",
      "94842866\tA\tG\t1/1\n",
      "94843657\tA\tG\t0|1\n",
      "94844516\tC\tG\t0|1\n",
      "94845570\tG\tC\t1/1\n",
      "94848458\tC\tT\t0/1\n",
      "94849307\tG\tA\t1/1\n",
      "94850760\tT\tC\t1/1\n",
      "94851443\tCaaataaataaataaataaataaataaataaataaataaataaataaataa\tCaaataaataaataaataaataaataaataaataaataaataaataa\t0|1\n",
      "94851871\tT\tC\t0|1\n",
      "94852077\tG\tA\t1/1\n",
      "94852283\tG\tA\t0|1\n",
      "94852614\tC\tT\t0|1\n",
      "94852716\tA\tG\t1|0\n"
     ]
    }
   ],
   "source": [
    "print(\"=== CYP2C19 variants ===\")\n",
    "!bcftools query -f '%POS\\t%REF\\t%ALT\\t[%GT]\\n' -r chr10:94759681-94853205 illumina_cyp_phased.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a6251438",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "=== CYP2C9 variants ===\n",
      "94938933\tT\tC\t0/1\n",
      "94939926\tG\tA\t0/1\n",
      "94940780\tC\tT\t1/1\n",
      "94940873\tA\tT\t1/1\n",
      "94941267\tgtttttttttttttttttttttttttt\tgTTtttttttttttttttttttttttttt\t1/1\n",
      "94941320\tG\tT\t0/1\n",
      "94942093\tT\tC\t0/1\n",
      "94943734\tC\tA\t0/1\n",
      "94944819\tA\tG\t1/1\n",
      "94945466\tA\tT\t0/1\n",
      "94946130\tT\tC\t1|0\n",
      "94946131\tG\tT\t1|0\n",
      "94946177\tA\tG\t1/1\n",
      "94946652\tC\tG\t0/1\n",
      "94950236\tT\tC\t1/1\n",
      "94950968\tC\tT\t1/1\n",
      "94952127\tG\tA\t0|1\n",
      "94952386\tT\tC\t1|0\n",
      "94952643\tA\tG\t1/1\n",
      "94952785\ttatagata\ttATAGatagata\t0|1\n",
      "94953443\tT\tC\t0/1\n",
      "94954003\tG\tA\t0/1\n",
      "94955914\tT\tC\t1|0\n",
      "94956377\tC\tT\t1|0\n",
      "94956617\tA\tC\t1|0\n",
      "94957439\tT\tG\t1/1\n",
      "94958456\tT\tC\t1|0\n",
      "94959037\tC\tT\t1|0\n",
      "94959052\tG\tT\t0|1\n",
      "94959431\tA\tG\t1/1\n",
      "94959719\tA\tG\t1/1\n",
      "94960070\tA\tG\t1/1\n",
      "94960275\tT\tC\t1/1\n",
      "94960761\tA\tG\t0/1\n",
      "94961350\tG\tC\t1/1\n",
      "94962302\tC\tA\t1/1\n",
      "94963667\tA\tG\t0/1\n",
      "94964413\tT\tC\t1/1\n",
      "94966538\ttcccc\ttccc\t0/1\n",
      "94967800\tG\tA\t1|0\n",
      "94968273\tC\tT\t1|0\n",
      "94968463\tA\tC\t1|0\n",
      "94968568\tT\tG\t1|0\n",
      "94968623\tgttttttttttt\tgtttttttttt\t0|1\n",
      "94969284\tA\tC\t0/1\n",
      "94970667\tG\tA\t1|0\n",
      "94971060\tT\tA\t0|1\n",
      "94971535\tC\tT\t1|0\n",
      "94972340\tA\tG\t0/1\n",
      "94972974\tT\tG\t1/1\n",
      "94973780\tT\tC\t1/1\n",
      "94974535\tgtttttttttttt\tgTTtttttttttttt,gTTTTtttttttttttt\t1/2\n",
      "94975298\tT\tG\t0/1\n",
      "94976289\tT\tC\t1|0\n",
      "94976351\tATTTTtttt\tATTTtttt\t0|1\n",
      "94976850\tA\tC\t1|0\n",
      "94977614\tC\tT\t1/1\n",
      "94978178\tT\tC\t1/1\n",
      "94978284\tC\tT\t0/1\n",
      "94979281\tgtttttttttttttttttttttttttt\tgttttttttttttttttttttttt\t1/1\n",
      "94981301\tC\tG\t1|0\n",
      "94981550\tT\tA\t1|0\n",
      "94982060\tA\tG\t1/1\n",
      "94982980\tC\tT\t1|0\n",
      "94982993\tC\tT\t1|0\n",
      "94984137\tA\tT\t0/1\n",
      "94984859\tC\tT\t0/1\n",
      "94988738\tA\tT\t1/1\n",
      "94989893\tT\tC\t1/1\n",
      "94990857\tT\tC\t1|0\n",
      "94991002\tG\tC\t1|0\n"
     ]
    }
   ],
   "source": [
    "print(\"\\n=== CYP2C9 variants ===\")\n",
    "!bcftools query -f '%POS\\t%REF\\t%ALT\\t[%GT]\\n' -r chr10:94938683-94993415 illumina_cyp_phased.vcf.gz\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3755efa6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "=== CYP2C8 variants ===\n",
      "94842866\tA\tG\t1/1\n",
      "94843657\tA\tG\t0|1\n",
      "94844516\tC\tG\t0|1\n",
      "94845570\tG\tC\t1/1\n",
      "94848458\tC\tT\t0/1\n",
      "94849307\tG\tA\t1/1\n",
      "94850760\tT\tC\t1/1\n",
      "94851443\tCaaataaataaataaataaataaataaataaataaataaataaataaataa\tCaaataaataaataaataaataaataaataaataaataaataaataa\t0|1\n",
      "94851871\tT\tC\t0|1\n",
      "94852077\tG\tA\t1/1\n",
      "94852283\tG\tA\t0|1\n",
      "94852614\tC\tT\t0|1\n",
      "94852716\tA\tG\t1|0\n",
      "94853547\tCtttttttttttttttt\tCTtttttttttttttttt\t1/1\n",
      "94854102\ttccacc\ttcc\t0/1\n",
      "94854968\tG\tT\t0/1\n",
      "94856167\tC\tT\t0/1\n",
      "94856660\tA\tT\t1/1\n",
      "94857274\tT\tG\t0/1\n",
      "94858114\tA\tG\t0/1\n",
      "94859474\tC\tT\t0/1\n",
      "94859761\tGCCCCCCCC\tGCCCCCCCCCC,GCCCCCCCCC\t1/2\n",
      "94859849\tG\tA\t1/1\n",
      "94860797\tA\tG\t1/1\n",
      "94860863\tcaaaaaaaaa\tcAaaaaaaaaa\t0/1\n",
      "94861449\tG\tC\t1/1\n",
      "94861942\tT\tC\t0/1\n",
      "94864151\tgtttttttttttt\tgTtttttttttttt\t0/1\n",
      "94866231\tC\tT\t1|0\n",
      "94866445\tT\tC\t0|1\n",
      "94866498\tG\tA\t0|1\n",
      "94868517\tT\tC\t0|1\n",
      "94868852\tC\tA\t0|1\n",
      "94869480\tA\tC\t1|0\n",
      "94870493\tcaaaaaaaaaaaaaa\tcaaaaaaaaaaaaa\t0|1\n",
      "94870503\tA\tT\t1|0\n",
      "94872496\tT\tC\t1/1\n",
      "94873074\tA\tG\t0/1\n",
      "94874084\tC\tT\t0/1\n",
      "94876348\tT\tC\t0/1\n",
      "94877910\tA\tG\t1/1\n",
      "94878357\tC\tT\t0/1\n",
      "94879614\tT\tA\t1/1\n",
      "94880345\tC\tT\t0/1\n",
      "94880506\tcttttttttttttt\tctttttttttttt\t0|1\n",
      "94880958\tT\tC\t1/1\n",
      "94882407\tG\tA\t0/1\n",
      "94883513\tcaaaaaaaaaaaa\tcAaaaaaaaaaaaa\t0|1\n",
      "94883514\taa\taACa\t0|1\n",
      "94884219\tG\tC\t1/1\n",
      "94885736\ttc\ttCc\t1/1\n",
      "94887046\tC\tA\t1/1\n",
      "94887360\tC\tT\t1/1\n",
      "94888280\tG\tA\t1/1\n",
      "94888467\tT\tG\t1/1\n",
      "94888512\tC\tT\t1|0\n",
      "94889074\tC\tT\t1/1\n",
      "94889097\tA\tT\t0|1\n",
      "94889643\tA\tC\t0|1\n",
      "94889866\tT\tC\t1/1\n",
      "94890131\tcttttttttttttttttt\tctttttttttttttt\t0|1\n",
      "94890436\tG\tA\t0|1\n",
      "94890571\tA\tG\t1/1\n",
      "94891201\tcaaaaaaaaaaaaaaa\tcAaaaaaaaaaaaaaaa\t0/1\n",
      "94891887\tG\tA\t0/1\n",
      "94892563\tG\tA\t0/1\n",
      "94893316\tG\tA\t1|0\n",
      "94893356\tcaaa\tcaa\t0|1\n",
      "94893362\tC\tT\t0|1\n",
      "94895202\tC\tT\t1/1\n",
      "94896031\tC\tG\t0|1\n",
      "94896123\tG\tA\t1|0\n",
      "94896161\tC\tT\t1/1\n",
      "94896626\tT\tC\t0|1\n",
      "94897200\tA\tG\t1|0\n",
      "94897664\tA\tG\t1|0\n",
      "94897715\tT\tG\t1/1\n",
      "94897850\tA\tG\t1/1\n",
      "94897981\tC\tT\t0|1\n",
      "94898147\tG\tC\t1|0\n",
      "94898443\tA\tT\t1|0\n",
      "94898738\tG\tA\t1|0\n",
      "94898897\tT\tC\t1/1\n",
      "94899091\tT\tG\t1/1\n",
      "94900374\tG\tT\t0|1\n",
      "94900915\tC\tG\t1|0\n",
      "94901052\tC\tT\t1|0\n",
      "94901551\tG\tC\t1/1\n",
      "94901662\tA\tC\t1/1\n",
      "94901882\tgc\tgCc\t1/1\n",
      "94901884\tA\tT\t1/1\n",
      "94901885\tA\tG\t1/1\n",
      "94901910\tcaaaaaaaaaaaaaaa\tcaaaaaaaaaaaaaa\t0/1\n",
      "94902312\tTAA\tTAAGAA\t1/1\n",
      "94902473\tC\tG\t0|1\n",
      "94902715\tC\tT\t1/1\n",
      "94903081\tA\tG\t0|1\n",
      "94903596\tG\tA\t1/1\n",
      "94903652\tG\tT\t1/1\n",
      "94904062\tA\tG\t1/1\n",
      "94905835\tG\tA\t0/1\n",
      "94906840\tG\tA\t1|0\n",
      "94906909\tT\tC\t0|1\n",
      "94907611\tG\tT\t1/1\n",
      "94908740\tggtgtg\tggtg\t0/1\n",
      "94909619\tC\tT\t0/1\n",
      "94910181\tT\tC\t1/1\n",
      "94912344\tC\tT\t0|1\n",
      "94912618\tC\tA\t0|1\n",
      "94913585\tG\tA\t0/1\n",
      "94914437\tcttttttttttttttt\tctttttttttttttt\t0|1\n",
      "94914899\tG\tA\t1|0\n",
      "94914964\tG\tA\t0|1\n",
      "94915966\tT\tC\t1/1\n",
      "94916470\tG\tA\t1/1\n",
      "94917537\tA\tG\t0/1\n",
      "94918214\tG\tT\t1/1\n",
      "94918531\tT\tA\t0|1\n",
      "94918842\tT\tC\t0|1\n",
      "94920508\tC\tT\t1/1\n",
      "94921243\tG\tA\t1|0\n",
      "94921282\tT\tC\t1|0\n",
      "94922312\tT\tC\t1|0\n",
      "94922558\tT\tA\t1/1\n",
      "94922572\tC\tT\t1|0\n",
      "94922757\tT\tC\t0|1\n",
      "94923082\tG\tA\t0|1\n",
      "94923793\tA\tT\t0|1\n",
      "94923794\tTTT\tTTTTGGTT\t0|1\n"
     ]
    }
   ],
   "source": [
    "print(\"\\n=== CYP2C8 variants ===\")\n",
    "!bcftools query -f '%POS\\t%REF\\t%ALT\\t[%GT]\\n' -r chr10:94842701-94924000 illumina_cyp_phased.vcf.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3256812b",
   "metadata": {},
   "source": [
    "### Star-Allele Calls\n",
    "\n",
    "**CYP2C19: likely *1/*1** because:\n",
    "- Did not find the major star-allele defining variants at their expected positions\n",
    "- CYP2C19*2 (rs4244285): not found\n",
    "- CYP2C19*17 (rs12248560): not found\n",
    "\n",
    "\n",
    "**CYP2C9: likely *1/*1** because:\n",
    "- Did NOT find the major star-allele defining variants\n",
    "- CYP2C9*2 (rs1799853): not found  \n",
    "- CYP2C9*3 (rs1057910): not found\n",
    "\n",
    "**CYP2C8: uncertain** because:\n",
    "- FOUND variant at position 94842866 (A>G, homozygous) matching rs11572080 (R139K), a CYP2C8*3 defining variant\n",
    "- Did NOT find the second *3 variant (rs10509681, K399R) at expected position\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6e7864d",
   "metadata": {},
   "source": [
    "# AI Usage Documentation\n",
    "\n",
    "\n",
    "## Summary of Usage\n",
    "Claude Sonnet 4.5 was used throughout this assignment for code generation and result interpretation. \n",
    "\n",
    "---\n",
    "\n",
    "## Detailed Usage by Section\n",
    "\n",
    "### 1. Reference Genome Download\n",
    "  - Commands for downloading chr10 from UCSC/Ensembl\n",
    "  - wget/curl syntax\n",
    "  - FASTA file handling\n",
    "\n",
    "### 2. Alignment (minimap2)\n",
    "  - minimap2 parameter selection for Illumina vs PacBio data\n",
    "  - Proper preset flags (-ax sr vs -ax map-pb)\n",
    "  - SAM to BAM conversion and indexing commands\n",
    "  \n",
    "\n",
    "### 3. Variant Calling (bcftools)\n",
    "  - bcftools mpileup and call command syntax\n",
    "  - Parameter selection \n",
    "  - VCF filtering and normalization\n",
    "  \n",
    "\n",
    "### 4. Phasing (HapCUT2/WhatsHap)\n",
    "  - Installation instructions for phasing tools\n",
    "  - Command syntax for HapCUT2 and WhatsHap\n",
    "  - Converting HapCUT2 block format to phased VCF\n",
    "  - Troubleshooting phasing errors\n",
    "\n",
    "### 5. Variant Analysis & Comparison\n",
    "  - bcftools isec commands for VCF comparison\n",
    "  - Extracting shared vs unique variants\n",
    "  - IGV automation scripting\n",
    "  - Interpreting IGV screenshots\n",
    "\n",
    "\n",
    "### 6. Star-Allele Determination\n",
    "  - Understanding PharmVar database structure\n",
    " \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (Anaconda base)",
   "language": "python",
   "name": "base"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
